politeiad/cmd/politeia_verify/politeia_verify.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2017 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/agl/ed25519"
+	"github.com/decred/dcrtime/merkle"
+)
+
+var (
+	publicKeyFlag = flag.String("k", "", "server public key")
+	tokenFlag     = flag.String("t", "", "proposal censorship token")
+	signatureFlag = flag.String("s", "", "proposal censorship signature")
+	verboseFlag   = flag.Bool("v", false, "verbose output")
+)
+
+func usage() {
+	fmt.Fprintf(os.Stderr, "usage: politeia_verify [-v] -k <pubkey> -t <token> -s <signature> <filename>...\n")
+	fmt.Fprintf(os.Stderr, " parameters:\n")
+	fmt.Fprintf(os.Stderr, "  pubkey        - Politiea's public server key\n")
+	fmt.Fprintf(os.Stderr, "  token         - Proposal censorship token\n")
+	fmt.Fprintf(os.Stderr, "  signature     - Proposal censorship "+
+		"signature\n")
+	fmt.Fprintf(os.Stderr, "  v             - Verbose output\n")
+	fmt.Fprintf(os.Stderr, "  filename      - One or more paths to the markdown "+
+		"and image files that make up the proposal\n")
+	fmt.Fprintf(os.Stderr, "\n")
+}
+
+func verifyProposal(key [ed25519.PublicKeySize]byte, token []byte, signature [ed25519.SignatureSize]byte) error {
+	flags := flag.Args()
+	if len(flags) < 1 {
+		usage()
+		return fmt.Errorf("must provide at least one filename for the proposal")
+	}
+
+	// Open all files and digest them.
+	hashes := make([]*[sha256.Size]byte, 0, len(flags))
+	for _, filename := range flags {
+		var payload []byte
+		payload, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return err
+		}
+
+		// Digest
+		h := sha256.New()
+		h.Write(payload)
+		digest := h.Sum(nil)
+
+		var digest32 [sha256.Size]byte
+		copy(digest32[:], digest)
+		hashes = append(hashes, &digest32)
+	}
+
+	merkle := *merkle.Root(hashes)
+	merkleToken := make([]byte, len(merkle)+len(token))
+	copy(merkleToken, merkle[:])
+	copy(merkleToken[len(merkle[:]):], token)
+	if ed25519.Verify(&key, merkleToken, &signature) {
+		if *verboseFlag {
+			fmt.Println("Proposal successfully verified.")
+		} else {
+			fmt.Println("0")
```

Comment:
I think you misunderstood me.  It is ok to print pertinent information when not-verbose.  So we should print the same message for both.  The difference in verbose is that you print additional information such as the merkles etc.  Printing 0 or 1 is certainly undesired.

Return err if it fails and that gets eventually translated into a printed error and 1 as the return value of the tool.

So make sure that you don't end up with double prints when erroring.

Reply:
@marcopeereboom updated the PR